paimon-core/src/main/java/org/apache/paimon/metrics/AbstractMetricGroup.java

```diff
@@ -0,0 +1,187 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.paimon.metrics;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Contains key functionality for adding metrics and carries metrics.
+ *
+ * <p>A MetricGroup can be {@link #close() closed}. Upon closing, the group de-register all metrics.
+ *
+ * <p>The {@link #close()} method and {@link #addMetric(String, Metric)} method won't be invoked by
+ * multiple threads at the same time, {@link #addMetric(String, Metric)} and {@link #getMetrics()}
+ * have multi-threads problems, so they should be synchronized by the object lock.
+ */
+public abstract class AbstractMetricGroup implements MetricGroup {
+    protected static final Logger LOG = LoggerFactory.getLogger(MetricGroup.class);
+
+    // ------------------------------------------------------------------------
+
+    /** The map containing all tags and their associated values, lazily computed. */
+    protected Map<String, String> tags;
+
+    /** Flag indicating whether this group has been closed. */
+    private boolean closed = false;
+
+    private final Map<String, Metric> metrics = new HashMap<>();
+
+    private final String table;
+
+    // ------------------------------------------------------------------------
+
+    public AbstractMetricGroup(String table, @Nullable Map<String, String> tags) {
+        this.table = table;
+        this.tags = tags;
+        Metrics.getInstance().addGroup(this);
+    }
+
+    @Override
+    public Map<String, String> getAllTags() {
+        return tags;
+    }
+
+    /**
+     * Returns the fully qualified metric name using the configured delimiter for the reporter with
+     * the given index, for example {@code "myTable.bucket-1.metricName"}.
+     *
+     * @param metricName metric name
+     * @param delimiter delimiter to use
+     * @return fully qualified metric name
+     */
+    public String getMetricIdentifier(String metricName, String delimiter) {
+        return String.join(delimiter, table, metricName);
+    }
+
+    /**
+     * Creates and registers a new {@link org.apache.paimon.metrics.Counter}.
+     *
+     * @param name name of the counter
+     * @return the created counter
+     */
+    public Counter counter(String name) {
+        return counter(name, new SimpleCounter());
```

Comment:
From a user's perspective, what I'm expecting is that, if I call `counter(name)` with the same name twice, I should get the same `Counter` object. Your implementation will throw away the counter created earlier.

Reply:
Make sense, add the handling for the metric name collision in `addMetric` method, which is a common metrics name collision handling for all the metric type. We keep the old metric and discard the new one with a warning message.